### PR TITLE
gh-112301: Enable warning emitting options and ignore warnings in CI

### PR DESCRIPTION
**Pull request by nohlson**
This PR introduces new compiler options that only emit warnings and don't affect runtime performance and some improvements to the warning check tooling.

Warning check tooling improvements:
- Clang output parser can pull out the option that triggered the warning
- A path prefix argument to the script that will be `lstrip()` from the file path parsed from the compiler output
- For each file with unexpected warnings a summary will be printed with the number of warnings found vs. expected
- For each file with unexpected improvements a summary will be printed with the number of warnings found vs. expected

Compiler options enabled:
- `-Wconversion`
- `-Wimplicit-fallthrough`
- `-Werror=format-security`
-  `-Wbidi-chars=any`
-  `-Wall`

This PR also modifies the reusable-macos and reusable-ubuntu jobs so that compiler output is only written to a file and warning checking is only performed on the specific jobs that we want:
- macos-14 non-free-threading
- ubuntu non-free-threading

All warnings have been added to the warning ignore files for the respective platforms.

This is one step in the [roadmap for compiler hardening](https://discuss.python.org/t/new-default-compiler-options-for-safety/60057/3).

* Issue: gh-112301

### CODE REVIEW

File: Misc/NEWS.d/next/Security/2024-08-06-00-06-23.gh-issue-112301.4k4lw6.rst
Comment (nohlson) [inline Misc/NEWS.d/next/Security/2024-08-06-00-06-23.gh-issue-112301.4k4lw6.rst]: This PR depends on already reviewed #122711 and https://github.com/python/devguide/pull/1364 to merge

File: .github/workflows/reusable-macos.yml
Comment (hugovk) [inline .github/workflows/reusable-macos.yml]:
```suggestion
    - name: Build CPython for compiler warning check
```

File: .github/workflows/reusable-macos.yml
Comment (hugovk) [inline .github/workflows/reusable-macos.yml]: Why `-j1` here? How long does it take compared with `-j4` or `-j8`?

File: .github/workflows/reusable-macos.yml
Comment (hugovk) [inline .github/workflows/reusable-macos.yml]: Why `-j4` here? It was `-j8` before.

File: Tools/build/check_warnings.py
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
    warnings in that file. Does not include duplicate warnings for a
```

File: Tools/build/check_warnings.py
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
    files_with_expected_warnings: set[FileWarnings],
```

File: Tools/build/check_warnings.py
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
    files_with_expected_warnings: set[FileWarnings],
```

File: configure.ac
Comment (erlend-aasland) [inline configure.ac:2517]: We should create a macro for reducing all this repetitive boilerplate code. Can be done in a follow-up PR.

File: .github/workflows/reusable-macos.yml
Comment (nohlson) [reply .github/workflows/reusable-macos.yml]: I found that with the macOS job there is a race condition between the threads to write out compiler output which would lead to malformed warnings. By using `-j1` this eliminates that issue

File: .github/workflows/reusable-macos.yml
Comment (nohlson) [reply .github/workflows/reusable-macos.yml]: I was experimenting with this option to see how it affected compiler output. The thread count in the Ubuntu job doesn't seem to impact the ability to write to stdout without malforming the diagnostics so I will revert this to `-j8`

Comment (nohlson) [reply .github/workflows/reusable-macos.yml]: Oh never mind this is for the non-warning-checking macOS job. Yes will definitely revert to `-j8`

File: Misc/NEWS.d/next/Security/2024-08-14-19-43-57.gh-issue-112301.IQUcOy.rst
Comment (sethmlarson) [inline Misc/NEWS.d/next/Security/2024-08-14-19-43-57.gh-issue-112301.IQUcOy.rst:1]: Should we list the compiler options that we enabled?

File: .github/workflows/reusable-macos.yml
Comment (nohlson) [reply .github/workflows/reusable-macos.yml]: @hugovk I pushed a commit that builds with `-j8` for the warning check build. You can see here that the output is malformed: https://github.com/python/cpython/actions/runs/10406271367/job/28818954178?pr=123020.

Comment (nohlson) [reply .github/workflows/reusable-macos.yml]: Turns out I can use home-brew in the macOS job to install make, then use gmake instead of the system BSD make to build. Then the option `--output-sync` can be enabled so that writing to stdout is thread safe. I put this in place and reverted to `-j8` and it looks like there are no issues in compiler output.

File: .github/workflows/reusable-macos.yml
Comment (hugovk) [inline .github/workflows/reusable-macos.yml]: Nit: let's move this step after the "Build CPython" step.

It moves it a little closer to "Check compiler warnings", so the two `if: ${{ !inputs.free-threading }} ...` ones are nearer each other and don't have the `if: ${{ inputs.free-threading }} ...` in between.

File: .github/workflows/reusable-ubuntu.yml
Comment (hugovk) [inline .github/workflows/reusable-ubuntu.yml]: Nit: same thing here: let's move this step after the "Build CPython out-of-tree" step.

File: .github/workflows/reusable-macos.yml
Comment (hugovk) [inline .github/workflows/reusable-macos.yml]: The CI didn't start:

> Invalid workflow file: .github/workflows/build.yml#L174
The workflow is not valid. In .github/workflows/build.yml (Line: 174, Col: 11): Error from called workflow python/cpython/.github/workflows/reusable-macos.yml@d7b53dbc48f0ec2597990fd7ff1677f5c6ab25e1 (Line: 51, Col: 12): The expression is not closed. An unescaped ${{ sequence was found, but the closing }} sequence was not found.


https://github.com/python/cpython/actions/runs/10408250107?pr=123020

Here we go:

```suggestion
      if : ${{ !inputs.free-threading && inputs.os == 'macos-13' }}
```

File: .github/workflows/reusable-macos.yml
Comment (hugovk) [reply .github/workflows/reusable-macos.yml]: Ah, right.

@erlend-aasland Do you think it's okay to use Homebrew's make here?

File: Tools/build/check_warnings.py
Comment (sethmlarson) [inline Tools/build/check_warnings.py]: `.lstrip()` doesn't work how you'd expect (instead removing characters passed individually, not as a unit), we should use `.removeprefix()` instead.

File: Tools/build/check_warnings.py
Comment (sethmlarson) [inline Tools/build/check_warnings.py]: How it's used now `path_prefix` can't be `None`, maybe should be `""` as a default?

File: Tools/build/check_warnings.py
Comment (sethmlarson) [inline Tools/build/check_warnings.py]: Same as above about `removeprefix()`

**Review by sethmlarson — COMMENTED** [review]
Some comments for you:

File: Misc/NEWS.d/next/Security/2024-08-14-19-43-57.gh-issue-112301.IQUcOy.rst
Comment (nohlson) [reply Misc/NEWS.d/next/Security/2024-08-14-19-43-57.gh-issue-112301.IQUcOy.rst:1]: I think this was brought up before and we decided not to in the news

File: .github/workflows/reusable-macos.yml
Comment (nohlson) [reply .github/workflows/reusable-macos.yml]: Rearranged step order

File: .github/workflows/reusable-ubuntu.yml
Comment (nohlson) [reply .github/workflows/reusable-ubuntu.yml]: Rearranged step order

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py]: Wow very interesting I read the docs and learned something! Changed to `.removeprefix()`

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py]: Switched to default to `""` which is better I agree

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py]: changed to `removeprefix()`

**Comment by nohlson** [comment]
I have made the requested changes; please review again

File: .github/workflows/reusable-macos.yml
Comment (erlend-aasland) [reply .github/workflows/reusable-macos.yml]: Why do we need Homebrew's `make`? It seems to me from the above discussion, it's to mitigate some race conditions. Would it not be better to fix those first, and then continue working on this PR?

Comment (nohlson) [reply .github/workflows/reusable-macos.yml]: The race condition is within make itself. When you build using multiple threads (`-j8`) each of those threads races to write the diagnostics of the build to stdout and  the make output itself is jumbled. GNU make has an option to sync the output from the threads so that it is legible `--output-sync`.

File: Tools/build/.warningignore_macos
Comment (hugovk) [inline Tools/build/.warningignore_macos]: nit:
```suggestion
# /path/to/file (number of warnings in file)
```

**Review by sethmlarson — COMMENTED** [review]
My comments are resolved, I'll leave a +1 review.

I spoke with @nohlson privately about potentially breaking up the addition of warnings into separate pull requests so they're only "active" while we're actively remediating the individual warnings. This would minimize the chances that core developers are going to "run into" this effort slightly.

After Nate dug into it, it turns out that almost all of the existing warnings come from `-Wconversion` and the other options produce far fewer warnings, some of which already have pull requests out for their remediation (like `-Wformat=2`).

For this reason I am okay proceeding with this PR as-is as the gains from splitting up the individual options seems smaller than I originally imagined, if other devs have concerns about this approach please leave them below.

File: Tools/build/.warningignore_macos
Comment (hugovk) [inline Tools/build/.warningignore_macos]: Does it matter line 6 begins with no space, and the following ones do? Same with `Tools/build/.warningignore_ubuntu`?

**Review by hugovk — COMMENTED** [review]
We can't merge this yet because the CI is failing:

https://github.com/python/cpython/actions/runs/10609035033/job/29404117175?pr=123020

How can we fix it?

File: Tools/build/.warningignore_macos
Comment (nohlson) [reply Tools/build/.warningignore_macos]: Added parenthesis

File: Tools/build/.warningignore_macos
Comment (nohlson) [reply Tools/build/.warningignore_macos]: Removed the leading spaces, missed them when translating the tool output to the warning ignore files

**Comment by nohlson** [comment]
> We can't merge this yet because the CI is failing:
> 
> https://github.com/python/cpython/actions/runs/10609035033/job/29404117175?pr=123020
> 
> How can we fix it?



> We can't merge this yet because the CI is failing:
> 
> https://github.com/python/cpython/actions/runs/10609035033/job/29404117175?pr=123020
> 
> How can we fix it?

The warning ignore files need to be updated immediately before merging. This is because any warnings that have merged to `main` since the last time `main` was pulled in will introduce warning diffs.

**Review by hugovk — APPROVED** [review]
Unsurprisingly there were new warnings when merging `main` in again, but the CI error message was useful: it told me which file had a different number of warnings, and linked to the devguide instructions on how to update.

Thanks @nohlson, let's merge!

**Pull request merged**
